### ISSUE TITLE
Prefix html class names with js- when used by JavaScript?

### ISSUE THREAD

**Issue by gravitystorm** [openstreetmap/openstreetmap-website#3801]
### Problem

When I'm refactoring, or when I'm reading pull requests, it's often unclear to me if particular html class names are still required. 

For example, in #3779 the CSS rules for `.loader` were removed, but the class was still left in the html code in the .erb templates in the PR. I had to dig around to find out that there is still JavaScript code that needs those classes to remain.

It's a particular problem because our JavaScript code is only lightly tested, so it's quite easy to break things by accidentally removing a class that's no longer required for CSS, without realising that it's also used by JavaScript.

### Description

I think that we should follow the convention to prefix all the classes used in our Javascript with `js-`. I've seen this done elsewhere, and recommended elsewhere too. Github has a nice explanation in their eslint rule that enforces this on their codebase:

https://github.com/github/eslint-plugin-github/blob/main/docs/rules/js-class-name.md

I'd like to hear any feedback from other developers who think this is either a good or bad idea, before I implement it. My intention would be to implement this across all of our JavaScript code, and then introduce an eslint rule like the one above to help enforce the convention.

**Comment by AntonKhorev** [comment]
> When I'm refactoring, or when I'm reading pull requests, it's often unclear to me if particular html class names are still required.

Search for class name. If they weren't removed in a pull request, they are likely still required.

> For example, in #3779 the CSS rules for `.loader` were removed, but the class was still left in the html code in the .erb templates in the PR. I had to dig around to find out that there is still JavaScript code that needs those classes to remain.

You can easily find them with a `.loader` search. `loader`/`loader_copy` classes still exist because there are about five different methods currently in use to place spinners. I want to replace them with one method (https://github.com/openstreetmap/openstreetmap-website/pull/3779#issuecomment-1317291481).

> I think that we should follow the convention to prefix all the classes used in our Javascript with `js-`.

Initially I wanted to say that since you don't like custom css, the only non-bootstrap classes should be the ones used in javascript. Then why don't you propose to prefix those temporary css-only classes instead? But then you wrote https://github.com/openstreetmap/openstreetmap-website/pull/3779#issuecomment-1318600407 so, I guess, javascript-used classes are also bad? Or are they only bad when added/removed?

> Github has a nice explanation in their eslint rule... "JavaScript should only query and handle events for js- prefixed class names."

Are they also against custom css?

So instead of this:

    <button type="button" class="btn-close" ...>

write this:

    <button type="button" class="btn-close js-btn-close" ...>

?

Do you want to prefix ids too? How something like this should look like

    $("#sidebar .search_form input[name=query]")

?

**Comment by tordans** [comment]
> I think that we should follow the convention to prefix all the classes used in our Javascript with js-. (…)
> I'd like to hear any feedback from other developers who think this is either a good or bad idea, before I implement it.

I agree that separating the names based on usage is a very good idea. 

In my last company, we had the rule to not query classes (or IDs) at all, but **use data attributes** instead. It would be `<div class="spinner" data-capability="spinner" />` with a lookup that also handled cases where multiple capabilities where attached (something like `data-capability~=spinner` I think).

The advantage is, that this splits the two works (style, javascript) even more.
However, IMO the advantage over `js-spinner` is mor stylistic.
